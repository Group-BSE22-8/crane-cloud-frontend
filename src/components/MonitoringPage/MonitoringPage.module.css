.MonitoringPageMain {
  flex-grow: 1;
  display: flex;
  flex-direction: column;
  background-color: #e5f3f9;
}

.PageContainer {
  max-width: 100%;
}

.PrimaryTitle {
  color: #008ac1;
  font-size: 52px;
  font-weight: bold;
  text-align: center;
  padding-top: 3rem;
}

.PrimaryContent,
.PrimaryDescription {
  color: #000000;
  font-size: 19px;
  font-weight: bold;
  text-align: center;
  padding-top: 1.5rem;
}

.BannerArea {
  display: flex;
  flex-wrap: wrap;
  text-align: center;
  justify-content: center;
  align-items: center;
  padding: 3rem 0 3rem 0;
}

.StatusBanner {
  background-color: #f7b21f;
  width: 35%;
}

.BannerText {
  padding: 1rem;
  color: #ffff;
  font-size: 1.5rem;
  text-transform: uppercase;
  font-weight: bold;
}

.StatusSectionTitle {
  padding: 1rem 0 0 5rem;
}

.StatusSection {
  display: grid;
  grid-template-columns: 1fr 1fr;
  grid-template-rows: 1fr 1fr;
  gap: 4rem;
}

.StatusContainer {
  display: flex;
  flex-direction: row;
  gap: 1rem;
}

.MoreDetailsIcon {
  padding: 1rem 1rem 2rem;
}

.MoreDetailsIcon svg {
  height: 24px;
}

.StatusContent {
  flex: 2;
}

.SmallIcon,
.SmallerIcon {
  width: 2rem;
  height: 2rem;
}

.SmallIcon,
.DownArrow,
.UpArrow {
  padding-top: 0.6rem;
}

.SmallIcon,
.DownArrow:hover,
.UpArrow:hover {
  cursor: pointer;
}

.StatusSectionContent {
  padding-left: 10rem;
  padding-bottom: 3rem;
  max-width: 70%;
}

.StatusSectionChildContainer {
  display: grid;
  grid-template-columns: 2fr 2fr;
  grid-template-rows: 1fr 1fr;
  gap: 1rem;
  padding-left: 5rem;
}

.StatusSectionItem {
  padding: 1rem;
  background-color: #ffff;
  color: #555555;
  display: grid;
  gap: 0.5rem;
}

.StatusSectionCardTitle {
  font-size: medium;
  line-height: 2rem;
  font-weight: bold;
}

.StatusSection {
  display: flex;
  flex-direction: row;
  justify-content: space-between;
}

.SelectWrapper {
  max-height: 6rem;
  display: grid;
  gap: 0.3rem;
}

.SelectOptionsWrapper {
  background-color: #ffffff;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  z-index: 3;
  box-shadow: 0px 5px 5px -3px rgba(0, 0, 0, 0.2),
    0px 8px 10px 1px rgba(0, 0, 0, 0.14), 0px 3px 14px 2px rgba(0, 0, 0, 0.12);
}

.DropDownContainer {
  background-color: #f9f9f9;
  position: absolute;
}

.DropDownContent {
  color: #000;
  background-color: #f9f9f9;
  display: flex;
  flex-direction: column;
  z-index: 1;
  box-shadow: 0 4px 4px 0 rgba(117, 117, 117, 0.2);
  -webkit-touch-callout: none;
  /* iOS Safari */
  -webkit-user-select: none;
  /* Safari */
  -khtml-user-select: none;
  /* Konqueror HTML */
  -moz-user-select: none;
  /* Old versions of Firefox */
  -ms-user-select: none;
  /* Internet Explorer/Edge */
  user-select: none;
  /* Non-prefixed version, currently supported by Chrome, Edge, Opera and Firefox */
}

.DropDownValue {
  display: flex;
  justify-content: space-between;
  gap: 3rem;
}

.DropDownValueTitle {
  font-weight: bold;
}

.DropDownDetails {
  cursor: pointer;
  padding: 1rem 3rem;
}

.DropDownDetails:hover {
  background-color: #f1f1f1;
}

.LoadingArea {
  padding: 4rem 0 4rem 0;
}

@media only screen and (max-width: 750px) {
  .StatusSectionContent {
    padding-left: 0rem;
  }

  .StatusSectionChildContainer {
    grid-template-columns: 1fr;
    grid-template-rows: 1fr;
    max-width: none;
  }

  .StatusTypes {
    display: none;
  }

  .StatusBanner {
    width: 80%;
  }
}
