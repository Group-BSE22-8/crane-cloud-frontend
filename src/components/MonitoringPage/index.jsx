import React, { useState, useEffect, useRef } from "react";
import NewHeader from "../NewHeader";
import LandingFooter from "../LandingFooter";
import { ReactComponent as Operational } from "../../assets/images/operational.svg";
import { ReactComponent as Maintenance } from "../../assets/images/maintenance.svg";
import { ReactComponent as Incident } from "../../assets/images/incident.svg";
import { ReactComponent as Outage } from "../../assets/images/off.svg";
import { ReactComponent as DownArrow } from "../../assets/images/down-arrow-black.svg";
import styles from "./MonitoringPage.module.css";

const MonitoringPage = () => {
  const [showOptions, setShowOptions] = useState(false);
  const openSelectRef = useRef(null);
  const toggleOptions = () => {
    setShowOptions(!showOptions);
  };
  const handleChange = () => {
    toggleOptions();
  };
  const handleClickOutside = (event) => {
    if (
      openSelectRef.current &&
      !openSelectRef.current.contains(event.target)
    ) {
      setShowOptions(false);
    }
  };
  useEffect(() => {
    document.addEventListener("mousedown", handleClickOutside);
    // returned function will be called on component unmount
    return () => {
      document.removeEventListener("mousedown", handleClickOutside);
    };
  }, []);
  return (
    <div className={styles.MonitoringPageMain}>
      <NewHeader />

      <div className={styles.PrimaryTitle}>Crane Cloud Status</div>
      <div className={styles.PrimaryContent}>
        Welcome to the Crane Cloud status page.
      </div>
      <div className={styles.PrimaryDescription}>
        Here you will find real-time information about the status of our
        services reported here.
      </div>
      <div className={styles.BannerArea}>
        <div className={styles.StatusBanner}>
          <div className={styles.BannerText}>All services operational</div>
        </div>
      </div>

      <div className={styles.StatusTypes}>
        <div className={styles.StatusTypesContainer}>
          <div className={styles.StatusTypeIcon}>
            <Operational />
          </div>
          <div className={styles.StatusTypeName}>
            <h3>Operational</h3>
          </div>
        </div>
        <div className={styles.StatusTypesContainer}>
          <div className={styles.StatusTypeIcon}>
            <Maintenance />
          </div>
          <div className={styles.StatusTypeName}>
            <h3>Maintenance</h3>
          </div>
        </div>
        <div className={styles.StatusTypesContainer}>
          <div className={styles.StatusTypeIcon}>
            <Incident />
          </div>
          <div className={styles.StatusTypeName}>
            <h3>Incident</h3>
          </div>
        </div>
        <div className={styles.StatusTypesContainer}>
          <div className={styles.StatusTypeIcon}>
            <Outage />
          </div>
          <div className={styles.StatusTypeName}>
            <h3>Outage</h3>
          </div>
        </div>
      </div>

      <div className={styles.StatusSectionTitle}>
        <h1>Current Status by Service</h1>
      </div>

      <div className={styles.StatusSectionContent}>
        <div className={styles.StatusSectionChildContainer}>
          <div className={styles.StatusSectionItem}>
            <div className={styles.StatusSection}>
              <span>
                <DownArrow className={styles.DownArrow} />
              </span>
              <span>
                <div className={styles.StatusSectionCardTitle}>Application</div>
                <div>No Issues</div>
              </span>
              <span>
                <Operational className={styles.SmallIcon} />
              </span>
            </div>
          </div>
          <div className={styles.StatusSectionItem}>
            <div className={styles.StatusSection}>
              <span>
                <div ref={openSelectRef} className="SelectWrapper">
                  <div onClick={() => toggleOptions()} role="presentation">
                    <div className={`${showOptions}`}>
                      <DownArrow className={styles.DownArrow} />
                    </div>
                  </div>
                  {showOptions && (
                    <div className="SelectOptionsWrapper">
                      <div onClick={() => handleChange()} role="presentation">
                        <div className={styles.StatusSectionItem}>
                          <div className={styles.StatusSection}>
                            <span>
                              <div className={styles.StatusSectionCardTitle}>
                                MySQL
                              </div>
                              <div>No Issues</div>
                            </span>
                            <span>
                              <Operational className={styles.SmallIcon} />
                            </span>
                          </div>
                        </div>
                        <hr />
                        <div className={styles.StatusSectionItem}>
                          <div className={styles.StatusSection}>
                            <span>
                              <div className={styles.StatusSectionCardTitle}>
                                PostgreSQL
                              </div>
                              <div>No Issues</div>
                            </span>
                            <span>
                              <Operational className={styles.SmallIcon} />
                            </span>
                          </div>
                        </div>
                      </div>
                    </div>
                  )}
                </div>
              </span>
              <span>
                <div className={styles.StatusSectionCardTitle}>
                  Database Servers
                </div>
                <div>No Issues</div>
              </span>
              <span>
                <Operational className={styles.SmallIcon} />
              </span>
            </div>
          </div>
          <div className={styles.StatusSectionItem}>
            <div className={styles.StatusSection}>
              <span>
                <DownArrow className={styles.DownArrow} />
              </span>
              <span>
                <div className={styles.StatusSectionCardTitle}>Clusters</div>
                <div>No Issues</div>
              </span>
              <span>
                <Operational className={styles.SmallIcon} />
              </span>
            </div>
          </div>
          <div className={styles.StatusSectionItem}>
            <div className={styles.StatusSection}>
              <span>
                <DownArrow className={styles.DownArrow} />
              </span>
              <span>
                <div className={styles.StatusSectionCardTitle}>Prometheus</div>
                <div>Issues detected</div>
              </span>
              <span>
                <Incident className={styles.SmallIcon} />
              </span>
            </div>
          </div>
          <div className={styles.StatusSectionItem}>
            <div className={styles.StatusSection}>
              <span>
                <DownArrow className={styles.DownArrow} />
              </span>
              <span>
                <div className={styles.StatusSectionCardTitle}>
                  Harbor Registry
                </div>
                <div>No Issues</div>
              </span>
              <span>
                <Operational className={styles.SmallIcon} />
              </span>
            </div>
          </div>
          <div className={styles.StatusSectionItem}>
            <div className={styles.StatusSection}>
              <span>
                <DownArrow className={styles.DownArrow} />
              </span>
              <span>
                <div className={styles.StatusSectionCardTitle}>MIRA</div>
                <div>No Issues</div>
              </span>
              <span>
                <Operational className={styles.SmallIcon} />
              </span>
            </div>
          </div>
          <div className={styles.StatusSectionItem}>
            <div className={styles.StatusSection}>
              <span>
                <DownArrow className={styles.DownArrow} />
              </span>
              <span>
                <div className={styles.StatusSectionCardTitle}>
                  Integrations
                </div>
                <div>No Issues</div>
              </span>
              <span>
                <Operational className={styles.SmallIcon} />
              </span>
            </div>
          </div>
        </div>
      </div>

      <div className={styles.FooterSection}>
        <LandingFooter />
      </div>
    </div>
  );
};

export default MonitoringPage;
