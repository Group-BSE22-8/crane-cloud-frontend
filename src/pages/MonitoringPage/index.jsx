import React, { useState, useEffect } from "react";
import NewHeader from "../../components/NewHeader";
import LandingFooter from "../../components/LandingFooter";
import { STATUS_MONITORING_URL } from "../../config";
import Spinner from "../../components/Spinner";
import { handleGetRequest } from "../../apis/apis";
import "./MonitoringPage.css";
import StatusGraph from "../../components/StatusGraph";
import StatusModule from "../../components/StatusModule";

const statusValue = [
  { type: "success" },
  { type: "partial" },
  { type: "failed" },
];

const MonitoringPage = () => {
  const [statusAvailable, setStatusAvailable] = useState("");
  const [statusData, setStatusData] = useState([]);
  const [loading, setLoading] = useState(false);

  const getStatusData = async () => {
    setLoading(true);
    try {
      await handleGetRequest(`${STATUS_MONITORING_URL}/statuses`).then((response) => {
        if (response.status !== 200) {
          return false;
        }
        setStatusAvailable(response.data.status);
        setStatusData(response.data.data);
        setLoading(false);
      });
    } catch (error) {
      setStatusData([]);
      setLoading(false);
    }
  };

  useEffect(() => {
    getStatusData();
  }, []);

  return (
    <div className="MonitoringPageMain">
      <NewHeader />

      <div className="PageContainer">
        <div className="PrimaryTitle">Crane Cloud Status</div>
        <div className="PrimaryContent">
          Welcome to the Crane Cloud status page.
        </div>
        <div className="PrimaryDescription">
          Here you will find real-time information about the status of our
          services reported here.
        </div>

        {statusAvailable === "success" ? (
          <>
            <div className="BannerArea">
              <div className="StatusBanner">
                {statusData?.cranecloud_status &&
                statusData?.database_status &&
                statusData?.mira_status &&
                statusData?.registry &&
                statusData?.clusters_status &&
                statusData?.prometheus_status === "success" ? (
                  <div className="BannerText">Services Available</div>
                ) : (
                  <div className="BannerText">Issues detected</div>
                )}
              </div>
            </div>

            <div className="StatusSectionContent">
              <div className="TitleArea">
                <div className="SectionTitle">Current Status Overview</div>
              </div>
              <div className="StatusSectionChildContainer">
                <StatusModule
                  title="Crane Cloud"
                  description="This includes the web application and platform API"
                  isOperational={
                    statusData?.cranecloud_status?.status ===
                    statusValue[0].type
                  }
                />

                <StatusModule
                  title="Databases"
                  description="This includes the MySQL and PostgreSQL database flavors"
                  isOperational={
                    statusData?.database_status?.status === statusValue[0].type
                  }
                />

                <StatusModule
                  title="Registry"
                  description="This is the Crane Cloud image registry"
                  isOperational={
                    statusData?.registry?.status === statusValue[0].type
                  }
                />

                <StatusModule
                  title="Mira Service"
                  description="This includes the Mira frontend and backend services"
                  isOperational={
                    statusData?.mira_status?.status === statusValue[0].type
                  }
                />

                <StatusModule
                  title="Clusters"
                  description="This includes the infrastructure Crane Cloud runs on"
                  isOperational={
                    statusData?.clusters_status?.status === statusValue[0].type
                  }
                />

                <StatusModule
                  title="Prometheus"
                  description="This is Crane Cloud's alerting toolkit"
                  isOperational={
                    statusData?.prometheus_status?.status ===
                    statusValue[0].type
                  }
                />
              </div>

              <div className="TitleArea">
                <div className="SectionTitle">System Status Series Graphs </div>
              </div>

              <StatusGraph />
            </div>
          </>
        ) : (
          loading && (
            <>
              <div className="LoadingArea">
                <Spinner size="big" />
              </div>
            </>
          )
        )}
      </div>
      <div className="FooterSection">
        <hr className="RowDivider" />
        <LandingFooter />
      </div>
    </div>
  );
};

export default MonitoringPage;
