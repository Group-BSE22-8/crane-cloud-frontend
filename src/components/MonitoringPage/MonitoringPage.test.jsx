import React from "react";
import { shallow } from "enzyme";
import Enzyme from "enzyme";
import MonitoringPage from "./index";

describe("Testing the MonitoringPage component", () => {
  const monitoringPageWrapper = shallow(<MonitoringPage />);
  const handleDropdown = jest.fn();
  const viewAppModules = jest.fn();
  const viewDBModules = jest.fn();
  const viewRegistryModules = jest.fn();
  const viewMiraModules = jest.fn();
  const handleClickOutside = jest.fn();
  const getStatusData = jest.fn();

  it("should match the monitoring page snapshot", () => {
    expect(monitoringPageWrapper).toMatchSnapshot();
    expect(handleDropdown).toBeDefined();
    expect(viewAppModules).toBeDefined();
    expect(viewDBModules).toBeDefined();
    expect(viewRegistryModules).toBeDefined();
    expect(viewMiraModules).toBeDefined();
    expect(handleClickOutside).toBeDefined();
    expect(getStatusData).toBeDefined();
  });

  it("assert the card dropdown arrow flip", () => {
    const ArrowFlipWrapper = monitoringPageWrapper
      .find(".CardArrow")
      .childAt(0);
    ArrowFlipWrapper.simulate("click");
    // next we verify that the arrow flips
    expect(monitoringPageWrapper.find(".CardArrow").childAt(0)).not.toEqual(
      ArrowFlipWrapper
    );
  });
});

describe("should check useState", () => {
  let wrapper;
  const setState = jest.fn();
  const useStateSpy = jest.spyOn(React, "useState");
  useStateSpy.mockImplementation((init) => [init, setState]);

  beforeEach(() => {
    wrapper = Enzyme.shallow(<MonitoringPage />);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  describe("Viewing App modules", () => {
    it("calls setOptionsFor", () => {
      wrapper.find(".CardArrow").props().onClick();
      expect(setState).toBeDefined();
    });
  });
});
