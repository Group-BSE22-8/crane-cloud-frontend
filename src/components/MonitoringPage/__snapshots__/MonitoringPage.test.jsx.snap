// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing the MonitoringPage component should match the monitoring page snapshot 1`] = `ShallowWrapper {}`;
